File: 0121-best-time-to-buy-and-sell-stock/javascript/Solution.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let n = prices.length
    let max = new Array(n)
    let curMax = 0
    for (let i = n-1; i >= 0; --i) {
        max[i] = curMax
        if (prices[i] > curMax) [
            curMax = prices[i]
        ]
    }
    let result = 0
    for (let i = 0; i < n; ++i) {
        if (result < max[i]-prices[i]) {
            result = max[i]-prices[i]
        }
    }
    return result
};

File: 0121-best-time-to-buy-and-sell-stock/cpp/Solution.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<int> max(n);
        int curMax = 0;
        for (int i = n - 1; i >= 0; --i) {
            max[i] = curMax;
            if (prices[i] > curMax) {
                curMax = prices[i];
            }
        }
        int result = 0;
        for (int i = 0; i < n; ++i) {
            if (result < max[i] - prices[i]) {
                result = max[i] - prices[i];
            }
        }
        return result;
    }
};
